_posts/2020-08-01-data-science-meets-devops.md

```diff
@@ -0,0 +1,769 @@
+
+
+<table>
+  <tr>
+   <td>
+<img src="/images/2020-08-01-data-science-meets-devops/fig3.reconciler.png" width="" alt="alt_text" title="image_tooltip">
+
+   </td>
+  </tr>
+  <tr>
+   <td>Figure 3. Illustration of the reconciler pattern as applied by our deployer.
+   </td>
+  </tr>
+</table>
+
+
+Reconcilers have proven immensely useful for building resilient systems because a well implemented reconciler provides a high degree of confidence that no matter how a system is perturbed it will eventually return to the desired state.
+
+
+### There Are No DAGs
+
+The declarative nature of controllers means data can flow through a series of controllers  without needing to explicitly create a DAG. In lieu of a DAG, a series of data processing steps can instead be expressed as a set of desired states, as illustrated in Figure 4 below:
+
+
+<table>
+  <tr>
+   <td>
+
+<img src="/images/2020-08-01-data-science-meets-devops/fig4.data-pipeline.png
+" width="" alt="alt_text" title="image_tooltip">
+
+   </td>
+  </tr>
+  <tr>
+   <td>Figure 4 illustrates how pipelines can emerge from independent controllers without explicitly encoding a DAG. Here we have two completely independent controllers. The first controller ensures that for every element a<sub>i</sub> there should be an element b<sub>i</sub>. The second controller ensures that for every element b<sub>i</sub> there should be an element c<sub>i</sub>. 
+   </td>
+  </tr>
+</table>
+
+
+This reconciler-based paradigm offers the following benefits over many traditional DAG-based workflows:
+
+
+
+*   **Resilience against failures**:  the system continuously seeks to achieve and maintain the desired state.  This state refers to both infrastructure (via Kubernetes manifests) and computation to be performed.  This is important in machine learning workflows where infrastructure is often strongly coupled with code (i.e. GPUs, databases, etc).
+*   **Increased autonomy of engineering teams:** each team is free to choose the tools and infrastructure that suit their needs.  The reconciler framework only requires a minimal amount of coupling between controllers while still allowing one to write expressive workflows.
+*   **Battle tested patterns and tools**:  This reconciler based framework does not invent something new.  Kubernetes has a rich ecosystem that makes building controllers super simple. The popularity of Kubernetes
```

Comment:
I would argue that building controllers is actually hard. Might be worth acknowledging this, and that it's getting easier over time.

Reply:
Reworded